File: src/Application Specific/Classes/Entity Collections/ContractsOriginal/ContractOriginalSite/ContractOriginalSiteSQL.vb
Imports System.Data.SqlClient

Namespace Entity

    Namespace ContractOriginalSites

        Public Class ContractOriginalSiteSQL
            Private _database As Sys.Database

            Public Sub New(ByVal database As Sys.Database)
                _database = database
            End Sub

#Region "Functions"

            Public Function [Get](ByVal ContractSiteID As Integer) As ContractOriginalSite
                _database.ClearParameter()
                _database.AddParameter("@ContractSiteID", ContractSiteID)

                'Get the datatable from the database store in dt
                Dim dt As DataTable = _database.ExecuteSP_DataTable("ContractOriginalSite_Get")

                If Not dt Is Nothing Then
                    If dt.Rows.Count > 0 Then
                        Dim oContractSite As New ContractOriginalSite
                        With oContractSite
                            .IgnoreExceptionsOnSetMethods = True
                            .SetContractSiteID = dt.Rows(0).Item("ContractSiteID")
                            .SetContractID = dt.Rows(0).Item("ContractID")
                            .SetPropertyID = dt.Rows(0).Item("SiteID")
                            .SetPricePerVisit = dt.Rows(0).Item("PricePerVisit")
                            .FirstVisitDate = CDate(dt.Rows(0).Item("FirstVisitDate"))
                            .SetVisitFrequencyID = dt.Rows(0).Item("VisitFrequencyID")
                            .SetVisitDuration = dt.Rows(0).Item("VisitDuration")
                            .SetAverageMileage = dt.Rows(0).Item("AverageMileage")
                            .SetIncludeAssetPrice = dt.Rows(0).Item("IncludeAssetPrice")
                            .SetIncludeMileagePrice = dt.Rows(0).Item("IncludeMileagePrice")
                            .SetPricePerMile = dt.Rows(0).Item("PricePerMile")
                            .SetTotalSitePrice = dt.Rows(0).Item("TotalSitePrice")
                            .SetIncludeRates = dt.Rows(0).Item("IncludeRates")
                            .SetLLCertReqd = dt.Rows(0).Item("LLCertReqd")
                            .SetAdditionalNotes = dt.Rows(0).Item("AdditionalNotes")
                            .SetCommercial = Entity.Sys.Helper.MakeBooleanValid(dt.Rows(0).Item("Commercial"))
                            .SetMainAppliances = (dt.Rows(0).Item("MainAppliances"))
                            .SetSecondryAppliances = (dt.Rows(0).Item("SecondryAppliances"))
                            .ContractSiteScheduleOfRates = DB.ContractOriginalSiteRates.ContractOriginalSiteRates_GetForContractSite(.ContractSiteID)
                        End With
                        oContractSite.Exists = True
                        Return oContractSite
                    Else
                        Return Nothing
                    End If
                Else
                    Return Nothing
                End If
            End Function

            Public Function GetAll() As DataView
                _database.ClearParameter()
                Dim dt As DataTable = _database.ExecuteSP_DataTable("ContractOriginalSite_GetAll")
                dt.TableName = Entity.Sys.Enums.TableNames.tblContractSite.ToString
                Return New DataView(dt)
            End Function

            Public Function GetAll_ContractID(ByVal ContractID As Integer, ByVal CustomerID As Integer) As DataView
                _database.ClearParameter()
                _database.AddParameter("@ContractID", ContractID, True)
                _database.AddParameter("@CustomerID", CustomerID, True)
                Dim dt As DataTable = _database.ExecuteSP_DataTable("ContractOriginalSite_GetAll_ContractID")
                dt.TableName = Entity.Sys.Enums.TableNames.tblContractSite.ToString
                Return New DataView(dt)
            End Function

            Public Function GetAll_ContractID2(ByVal ContractID As Integer, ByVal CustomerID As Integer) As DataView
                _database.ClearParameter()
                _database.AddParameter("@ContractID", ContractID, True)
                Dim dt As DataTable = _database.ExecuteSP_DataTable("ContractOriginalSite_GetAll_ContractID2")
                dt.TableName = Entity.Sys.Enums.TableNames.tblContractSite.ToString
                Return New DataView(dt)
            End Function

            Private Function GetContractOriginalSiteScheduleOfRate(ByVal ContractSiteID As Integer)
                _database.ClearParameter()
                _database.AddParameter("@ContractSiteID", ContractSiteID, True)
                Dim dt As DataTable = _database.ExecuteSP_DataTable("ContractOriginalSiteScheduleOfRate_Get")
                dt.TableName = Entity.Sys.Enums.TableNames.tblSiteScheduleOfRate.ToString
                Return New DataView(dt)
            End Function

            Public Function Insert(ByVal oContractSite As ContractOriginalSite) As ContractOriginalSite
                _database.ClearParameter()
                AddContractSiteParametersToCommand(oContractSite)

                oContractSite.SetContractSiteID = Entity.Sys.Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("ContractOriginalSite_Insert"))
                oContractSite.Exists = True
                'SaveRates(oContractSite)
                Return oContractSite
            End Function

            Public Sub Update(ByVal oContractSite As ContractOriginalSite)
                _database.ClearParameter()
                _database.AddParameter("@ContractSiteID", oContractSite.ContractSiteID, True)
                AddContractSiteParametersToCommand(oContractSite)
                _database.ExecuteSP_NO_Return("ContractOriginalSite_Update")
                'SaveRates(oContractSite)
            End Sub

            Public Function Delete(ByVal ContractSiteID As Integer) As Integer
                _database.ClearParameter()
                _database.AddParameter("@ContractSiteID", ContractSiteID, True)
                _database.AddParameter("@DownloadStatus", CInt(Entity.Sys.Enums.VisitStatus.Downloaded), True)
                Return _database.ExecuteSP_OBJECT("ContractOriginalSite_Delete")
            End Function

            Public Function ActiveInactive(ByVal ContractSiteID As Integer, ByVal InactiveFlag As Boolean) As Integer
                _database.ClearParameter()
                _database.AddParameter("@ContractSiteID", ContractSiteID, True)
                _database.AddParameter("@DownloadStatus", CInt(Entity.Sys.Enums.VisitStatus.Scheduled), True)
                _database.AddParameter("@InactiveFlag", InactiveFlag, True)
                Return _database.ExecuteSP_OBJECT("ContractOriginalSite_ActiveInactive")
            End Function

            Public Sub Delete_Visits(ByVal ContractSiteID As Integer)
                _database.ClearParameter()
                _database.AddParameter("@ContractSiteID", ContractSiteID, True)
                _database.ExecuteSP_OBJECT("ContractOriginalSite_Visits_Delete")
            End Sub

            Private Sub AddContractSiteParametersToCommand(ByRef oContractSite As ContractOriginalSite)
                With _database
                    .AddParameter("@ContractID", oContractSite.ContractID, True)
                    .AddParameter("@SiteID", oContractSite.PropertyID, True)
                    .AddParameter("@PricePerVisit", oContractSite.PricePerVisit, True)
                    .AddParameter("@FirstVisitDate", oContractSite.FirstVisitDate, True)
                    .AddParameter("@VisitFrequencyID", oContractSite.VisitFrequencyID, True)
                    .AddParameter("@VisitDuration", oContractSite.VisitDuration, True)
                    .AddParameter("@AverageMileage", oContractSite.AverageMileage, True)
                    .AddParameter("@IncludeAssetPrice", oContractSite.IncludeAssetPrice, True)
                    .AddParameter("@IncludeMileagePrice", oContractSite.IncludeMileagePrice, True)
                    .AddParameter("@PricePerMile", oContractSite.PricePerMile, True)
                    .AddParameter("@TotalSitePrice", oContractSite.TotalSitePrice, True)
                    .AddParameter("@IncludeRates", oContractSite.IncludeRates, True)
                    .AddParameter("@LLCertReqd", oContractSite.LLCertReqd, True)
                    .AddParameter("@AdditionalNotes", oContractSite.AdditionalNotes, True)
                    .AddParameter("@Commercial", oContractSite.Commercial, True)
                    .AddParameter("@MainAppliances", oContractSite.MainAppliances, True)
                    .AddParameter("@SecondryAppliances", oContractSite.SecondryAppliances, True)
                End With
            End Sub

            Private Sub SaveRates(ByVal oContractSite As ContractOriginalSite)
                _database.ClearParameter()
                _database.AddParameter("@ContractSiteID", oContractSite.ContractSiteID, True)
                _database.ExecuteSP_NO_Return("ContractOriginalSiteScheduleOfRate_Delete")

                If Not oContractSite.ContractSiteScheduleOfRates Is Nothing Then
                    For Each r As DataRow In oContractSite.ContractSiteScheduleOfRates.Table.Rows
                        _database.ClearParameter()
                        _database.AddParameter("@ContractSiteID", oContractSite.ContractSiteID, True)
                        _database.AddParameter("@ScheduleOfRatesCategoryID", r("ScheduleOfRatesCategoryID"), True)
                        _database.AddParameter("@Code", r("Code"), True)
                        _database.AddParameter("@Description", r("Description"), True)
                        _database.AddParameter("@Price", r("Price"), True)
                        _database.AddParameter("@QtyPerVisit", r("QtyPerVisit"), True)
                        _database.ExecuteSP_NO_Return("ContractOriginalSiteScheduleOfRate_Insert")

                    Next
                End If

            End Sub

#End Region

        End Class

    End Namespace

End Namespace

File: src/Application Specific/Classes/Entity Collections/ContractsOriginal/ContractOriginalSite/ContractOriginalSiteSQL.cs
using System.Data;
using Microsoft.VisualBasic.CompilerServices;

namespace FSM.Entity
{
    namespace ContractOriginalSites
    {
        public class ContractOriginalSiteSQL
        {
            private Sys.Database _database;

            public ContractOriginalSiteSQL(Sys.Database database)
            {
                _database = database;
            }

            public ContractOriginalSite Get(int ContractSiteID)
            {
                _database.ClearParameter();
                _database.AddParameter("@ContractSiteID", ContractSiteID);

                // Get the datatable from the database store in dt
                var dt = _database.ExecuteSP_DataTable("ContractOriginalSite_Get");
                if (dt is object)
                {
                    if (dt.Rows.Count > 0)
                    {
                        var oContractSite = new ContractOriginalSite();
                        oContractSite.IgnoreExceptionsOnSetMethods = true;
                        oContractSite.SetContractSiteID = dt.Rows[0]["ContractSiteID"];
                        oContractSite.SetContractID = dt.Rows[0]["ContractID"];
                        oContractSite.SetPropertyID = dt.Rows[0]["SiteID"];
                        oContractSite.SetPricePerVisit = dt.Rows[0]["PricePerVisit"];
                        oContractSite.FirstVisitDate = Conversions.ToDate(dt.Rows[0]["FirstVisitDate"]);
                        oContractSite.SetVisitFrequencyID = dt.Rows[0]["VisitFrequencyID"];
                        oContractSite.SetVisitDuration = dt.Rows[0]["VisitDuration"];
                        oContractSite.SetAverageMileage = dt.Rows[0]["AverageMileage"];
                        oContractSite.SetIncludeAssetPrice = dt.Rows[0]["IncludeAssetPrice"];
                        oContractSite.SetIncludeMileagePrice = dt.Rows[0]["IncludeMileagePrice"];
                        oContractSite.SetPricePerMile = dt.Rows[0]["PricePerMile"];
                        oContractSite.SetTotalSitePrice = dt.Rows[0]["TotalSitePrice"];
                        oContractSite.SetIncludeRates = dt.Rows[0]["IncludeRates"];
                        oContractSite.SetLLCertReqd = dt.Rows[0]["LLCertReqd"];
                        oContractSite.SetAdditionalNotes = dt.Rows[0]["AdditionalNotes"];
                        oContractSite.SetCommercial = Sys.Helper.MakeBooleanValid(dt.Rows[0]["Commercial"]);
                        oContractSite.SetMainAppliances = dt.Rows[0]["MainAppliances"];
                        oContractSite.SetSecondryAppliances = dt.Rows[0]["SecondryAppliances"];
                        oContractSite.ContractSiteScheduleOfRates = App.DB.ContractOriginalSiteRates.ContractOriginalSiteRates_GetForContractSite(oContractSite.ContractSiteID);
                        oContractSite.Exists = true;
                        return oContractSite;
                    }
                    else
                    {
                        return null;
                    }
                }
                else
                {
                    return null;
                }
            }

            public DataView GetAll()
            {
                _database.ClearParameter();
                var dt = _database.ExecuteSP_DataTable("ContractOriginalSite_GetAll");
                dt.TableName = Sys.Enums.TableNames.tblContractSite.ToString();
                return new DataView(dt);
            }

            public DataView GetAll_ContractID(int ContractID, int CustomerID)
            {
                _database.ClearParameter();
                _database.AddParameter("@ContractID", ContractID, true);
                _database.AddParameter("@CustomerID", CustomerID, true);
                var dt = _database.ExecuteSP_DataTable("ContractOriginalSite_GetAll_ContractID");
                dt.TableName = Sys.Enums.TableNames.tblContractSite.ToString();
                return new DataView(dt);
            }

            public DataView GetAll_ContractID2(int ContractID, int CustomerID)
            {
                _database.ClearParameter();
                _database.AddParameter("@ContractID", ContractID, true);
                var dt = _database.ExecuteSP_DataTable("ContractOriginalSite_GetAll_ContractID2");
                dt.TableName = Sys.Enums.TableNames.tblContractSite.ToString();
                return new DataView(dt);
            }

            private object GetContractOriginalSiteScheduleOfRate(int ContractSiteID)
            {
                _database.ClearParameter();
                _database.AddParameter("@ContractSiteID", ContractSiteID, true);
                var dt = _database.ExecuteSP_DataTable("ContractOriginalSiteScheduleOfRate_Get");
                dt.TableName = Sys.Enums.TableNames.tblSiteScheduleOfRate.ToString();
                return new DataView(dt);
            }

            public ContractOriginalSite Insert(ContractOriginalSite oContractSite)
            {
                _database.ClearParameter();
                AddContractSiteParametersToCommand(ref oContractSite);
                oContractSite.SetContractSiteID = Sys.Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("ContractOriginalSite_Insert"));
                oContractSite.Exists = true;
                // SaveRates(oContractSite)
                return oContractSite;
            }

            public void Update(ContractOriginalSite oContractSite)
            {
                _database.ClearParameter();
                _database.AddParameter("@ContractSiteID", oContractSite.ContractSiteID, true);
                AddContractSiteParametersToCommand(ref oContractSite);
                _database.ExecuteSP_NO_Return("ContractOriginalSite_Update");
                // SaveRates(oContractSite)
            }

            public int Delete(int ContractSiteID)
            {
                _database.ClearParameter();
                _database.AddParameter("@ContractSiteID", ContractSiteID, true);
                _database.AddParameter("@DownloadStatus", Conversions.ToInteger(Sys.Enums.VisitStatus.Downloaded), true);
                return Conversions.ToInteger(_database.ExecuteSP_OBJECT("ContractOriginalSite_Delete"));
            }

            public int ActiveInactive(int ContractSiteID, bool InactiveFlag)
            {
                _database.ClearParameter();
                _database.AddParameter("@ContractSiteID", ContractSiteID, true);
                _database.AddParameter("@DownloadStatus", Conversions.ToInteger(Sys.Enums.VisitStatus.Scheduled), true);
                _database.AddParameter("@InactiveFlag", InactiveFlag, true);
                return Conversions.ToInteger(_database.ExecuteSP_OBJECT("ContractOriginalSite_ActiveInactive"));
            }

            public void Delete_Visits(int ContractSiteID)
            {
                _database.ClearParameter();
                _database.AddParameter("@ContractSiteID", ContractSiteID, true);
                _database.ExecuteSP_OBJECT("ContractOriginalSite_Visits_Delete");
            }

            private void AddContractSiteParametersToCommand(ref ContractOriginalSite oContractSite)
            {
                {
                    var withBlock = _database;
                    withBlock.AddParameter("@ContractID", oContractSite.ContractID, true);
                    withBlock.AddParameter("@SiteID", oContractSite.PropertyID, true);
                    withBlock.AddParameter("@PricePerVisit", oContractSite.PricePerVisit, true);
                    withBlock.AddParameter("@FirstVisitDate", oContractSite.FirstVisitDate, true);
                    withBlock.AddParameter("@VisitFrequencyID", oContractSite.VisitFrequencyID, true);
                    withBlock.AddParameter("@VisitDuration", oContractSite.VisitDuration, true);
                    withBlock.AddParameter("@AverageMileage", oContractSite.AverageMileage, true);
                    withBlock.AddParameter("@IncludeAssetPrice", oContractSite.IncludeAssetPrice, true);
                    withBlock.AddParameter("@IncludeMileagePrice", oContractSite.IncludeMileagePrice, true);
                    withBlock.AddParameter("@PricePerMile", oContractSite.PricePerMile, true);
                    withBlock.AddParameter("@TotalSitePrice", oContractSite.TotalSitePrice, true);
                    withBlock.AddParameter("@IncludeRates", oContractSite.IncludeRates, true);
                    withBlock.AddParameter("@LLCertReqd", oContractSite.LLCertReqd, true);
                    withBlock.AddParameter("@AdditionalNotes", oContractSite.AdditionalNotes, true);
                    withBlock.AddParameter("@Commercial", oContractSite.Commercial, true);
                    withBlock.AddParameter("@MainAppliances", oContractSite.MainAppliances, true);
                    withBlock.AddParameter("@SecondryAppliances", oContractSite.SecondryAppliances, true);
                }
            }
        }
    }
}